src/Functions/hilbertDecode.h

```diff
@@ -0,0 +1,221 @@
+    static std::pair<UInt64, UInt64> getCodeAndUpdateState(UInt8 hilbert_bits, UInt8& state)
+    {
+        const UInt8 table_index = state | hilbert_bits;
+        const auto table_code = HilbertDetails::HilbertDecodeLookupTable<bit_step>::LOOKUP_TABLE[table_index];
+        state = table_code & STATE_MASK;
+        const UInt64 x_bits = (table_code & X_MASK) >> bit_step;
+        const UInt64 y_bits = table_code & Y_MASK;
+        return {x_bits, y_bits};
+    }
+
+    // hilbert code is double size of input values
+    static constexpr UInt8 getHilbertShift(UInt8 shift)
+    {
+        return shift << 1;
+    }
+
+    static std::pair<Int8, UInt8> getInitialShiftAndState(UInt8 used_bits)
+    {
+        UInt8 iterations = used_bits / HILBERT_SHIFT;
+        Int8 initial_shift = iterations * HILBERT_SHIFT;
+        if (initial_shift < used_bits)
+        {
+            ++iterations;
+        }
+        else
+        {
+            initial_shift -= HILBERT_SHIFT;
+        }
+        UInt8 state = iterations % 2 == 0 ? LEFT_STATE : DEFAULT_STATE;
+        return {initial_shift, state};
+    }
+
+    constexpr static UInt8 STEP_MASK = (1 << bit_step) - 1;
+    constexpr static UInt8 HILBERT_SHIFT = getHilbertShift(bit_step);
+    constexpr static UInt8 HILBERT_MASK = (1 << HILBERT_SHIFT) - 1;
+    constexpr static UInt8 STATE_MASK = 0b11 << HILBERT_SHIFT;
+    constexpr static UInt8 Y_MASK = STEP_MASK;
+    constexpr static UInt8 X_MASK = STEP_MASK << bit_step;
+    constexpr static UInt8 LEFT_STATE = 0b01 << HILBERT_SHIFT;
+    constexpr static UInt8 DEFAULT_STATE = bit_step % 2 == 0 ? LEFT_STATE : 0;
+};
+
+
+class FunctionHilbertDecode : public FunctionSpaceFillingCurveDecode<2, 0, 32>
+{
+public:
+    static constexpr auto name = "hilbertDecode";
+    static FunctionPtr create(ContextPtr)
+    {
+        return std::make_shared<FunctionHilbertDecode>();
+    }
+
+    String getName() const override { return name; }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t input_rows_count) const override
+    {
+        size_t nd;
+        const auto * col_const = typeid_cast<const ColumnConst *>(arguments[0].column.get());
+        const auto * mask = typeid_cast<const ColumnTuple *>(col_const->getDataColumnPtr().get());
+        if (mask)
+            nd = mask->tupleSize();
+        else
+            nd = col_const->getUInt(0);
+        auto non_const_arguments = arguments;
+        non_const_arguments[1].column = non_const_arguments[1].column->convertToFullColumnIfConst();
+        const ColumnPtr & col_code = non_const_arguments[1].column;
+        Columns tuple_columns(nd);
+
+        const auto shrink = [mask](const UInt64 value, const UInt8 column_id)
+        {
+            if (mask)
+                return value >> mask->getColumn(column_id).getUInt(0);
+            return value;
+        };
+
+        auto col0 = ColumnUInt64::create();
+        auto & vec0 = col0->getData();
```

Comment:
also, we can add more representative variable names in these two cases

Reply:
hm, it will be just first component of result vector (aka x-axis), (like hilbertDecode(2, 31) = (3, 4)) so i dont know what other name will be suitable here (we can name x = col0, y = col1, but it will be bad when more dimensions will be added)